File: 2354-number-of-excellent-pairs/2354-number-of-excellent-pairs.cpp
class Solution {
public:
 long long countExcellentPairs(vector<int>& A, int k) {
    long long bits[30] = {}, ans = 0;
    for (int n : unordered_set<int>(begin(A), end(A)))
        ++bits[__builtin_popcount(n)];
    for (int i = 1; i < 30; ++i)
        for (int j = 1; j < 30; ++j)
            if (i + j >= k)
                ans += bits[i] * bits[j];
    return ans;
 }
};

File: 2354-number-of-excellent-pairs/2354-number-of-excellent-pairs.py
class Solution:
    def countExcellentPairs(self, nums: List[int], k: int) -> int:
        distinct_elements = set()
        for num in nums:
            distinct_elements.add(num)
         
        def countOnes (num):
            count = 0
            while(num):
                count += 1
                num &= num - 1
            return count
        
        bits = []
        
        for element in distinct_elements:
            bits.append(countOnes(element))
            
        bits.sort()
        answer = 0
        
        for bit in bits:
            firstBit = bisect.bisect_left(bits, k - bit, lo=0, hi=len(bits)) 
            answer += len(bits) - firstBit
            
        return answer
        
       
        